packages/netlify-cms-backend-github/src/GraphQLAPI.js

```diff
@@ -0,0 +1,459 @@
+      });
+      // https://developer.github.com/v4/enum/repositorypermission/
+      const { viewerPermission } = data.repository;
+      return ['ADMIN', 'MAINTAIN', 'WRITE'].includes(viewerPermission);
+    } catch (error) {
+      console.error('Problem fetching repo data from GitHub');
+      throw error;
+    }
+  }
+
+  async user() {
+    const { data } = await this.query({
+      query: queries.user,
+      fetchPolicy: CACHE_FIRST, // we can assume user details don't change often
+    });
+    return data.viewer;
+  }
+
+  async retrieveBlobObject(expression, options = {}) {
+    const { repo_owner: owner, repo_name: name } = this;
+    const { data } = await this.query({
+      query: queries.blob,
+      variables: { owner, name, expression },
+      ...options,
+    });
+    // https://developer.github.com/v4/object/blob/
+    if (data.repository.object) {
+      const { is_binary, text } = data.repository.object;
+      return { is_null: false, is_binary, text };
+    } else {
+      return { is_null: true };
+    }
+  }
+
+  async retrieveContent(path, branch) {
+    const { is_null, is_binary, text } = await this.retrieveBlobObject(`${branch}:${path}`);
+    if (is_null) {
+      throw new APIError('Not Found', 404, 'GitHub');
+    } else if (!is_binary) {
+      return text;
+    } else {
+      return super.retrieveContent(path, branch);
+    }
+  }
+
+  async retrieveBlob(sha) {
+    const { is_null, is_binary, text } = await this.retrieveBlobObject(
+      sha,
+      { fetchPolicy: CACHE_FIRST }, // blob sha is derived from file content
+    );
+
+    if (is_null) {
+      throw new APIError('Not Found', 404, 'GitHub');
+    } else if (!is_binary) {
+      return text;
+    } else {
+      return super.retrieveBlob(sha);
+    }
+  }
+
+  async getStatuses(sha) {
+    const { repo_owner: owner, repo_name: name } = this;
+    const { data } = await this.query({ query: queries.statues, variables: { owner, name, sha } });
+    if (data.repository.object) {
+      const { status } = data.repository.object;
+      const { contexts } = status || { contexts: [] };
+      return contexts;
+    } else {
+      throw new APIError(`No commit found for SHA: ${sha}`, 422, 'GitHub');
+    }
+  }
+
+  async listFiles(path) {
+    const { repo_owner: owner, repo_name: name } = this;
+    const { data } = await this.query({
+      query: queries.files,
+      variables: { owner, name, expression: `${this.branch}:${path}` },
+    });
+
+    if (data.repository.object) {
+      const files = data.repository.object.entries.map(e => ({
+        ...e,
+        path: `${path}/${e.name}`,
+        download_url: `https://raw.githubusercontent.com/${this.repo}/${this.branch}/${path}/${e.name}`,
+        size: e.blob && e.blob.size,
+      }));
+      return files;
+    } else {
+      throw new APIError('Not Found', 404, 'GitHub');
+    }
+  }
+
+  async listUnpublishedBranches() {
```

Comment:
This single query replaces a much more complicated one in the base class (that has multiple requests)